README.md
# swipe_button package for Flutter

[![pub package](https://img.shields.io/pub/v/swipe_button.svg)](https://pub.dartlang.org/packages/swipe_button)

A Flutter package for iOS and Android for picking location and images.
## Demo
<img src="http://sheikhsoft.com/screensort/swipe_button.png" width="340" height="640" title="Screen Shoot">


## Installation

First, add `swipe_button` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

### iOS

No configuration required - the plugin should work out of the box.


### Android

No configuration required - the plugin should work out of the box.

### Code for the swipe_button widget

``` dart
import 'package:swipe_button/swipe_button.dart';

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  final swipeController = SwipeController();
 
  @override
  Widget build(BuildContext context) {
    return Container(
                         alignment: Alignment.center,
                         child: Padding(
                           padding: const EdgeInsets.all(24.0),
                           child: SwipeButton(
                             swipeController: swipeController,
                             thumb: Row(
                               mainAxisAlignment: MainAxisAlignment.center,
                               children: <Widget>[
                                 Align(
                                     widthFactor: 0.90,
                                     child: Icon(
                                       Icons.chevron_right,
                                       size: 60.0,
                                       color: Colors.white,
                                     )),
                               ],
                             ),
                             content: Center(
                               child: Text(
                                 buttonTitle,
                                 style: TextStyle(color: Colors.white),
                               ),
                             ),
                             onChanged: (result) {
                               if (result == SwipePosition.SwipeRight) {
                                 Navigator.pushNamed(context, pageRoute);
                               } else {
                                 // optional
                                 //swipeController.reset();
                               }
                             },
                           ),
                         ),
                       );
  }
  
}
```

### Code for the Full page swipe_button widget

``` dart
import 'package:swipe_button/swipe_button_demo.dart';

class HomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SwipeHomeWidget(
        pageRought: "/second",
      ),
    );
  }
}
```

CHANGELOG.md
## [0.0.2]
* `SwipeController` was added , it allows us to reset the position of the `SwipeButton`.

## [0.0.1]
* Initial release


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:swipe_button/swipe_button.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final swipeController = SwipeController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        actions: [
          IconButton(
            icon: Icon(Icons.restore),
            onPressed: () => swipeController.reset(),
          )
        ],
      ),
      body: Container(
        alignment: Alignment.center,
        child: Padding(
          padding: const EdgeInsets.all(24.0),
          child: SwipeButton(
            swipeController: swipeController,
            thumb: Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                Align(
                    widthFactor: 0.90,
                    child: Icon(
                      Icons.chevron_right,
                      size: 60.0,
                      color: Colors.white,
                    )),
              ],
            ),
            content: Center(
              child: Text(
                'Testing',
                style: TextStyle(color: Colors.white),
              ),
            ),
            onChanged: (result) {
              print(result);
              if (result == SwipePosition.SwipeRight) {
              } else {}
            },
          ),
        ),
      ),
    );
  }
}


lib/swipe_button.dart
library swipe_button;

import 'package:flutter/material.dart';
import 'package:flutter/physics.dart';

enum SwipePosition {
  SwipeLeft,
  SwipeRight,
}

/// Controller which allow us to reset the swipe button.
class SwipeController {
  AnimationController _controller;

  void _attachController(AnimationController controller) =>
      _controller = controller;

  /// Reset the Swipe button using a smooth animation
  /// Set the duration what you want
  void reset({Duration duration = const Duration(milliseconds: 600)}) {
    assert(_controller != null,
        'SwipeController was not attached to the SwipeButton');
    _controller.duration = duration;
    _controller.reverse();
  }
}

class SwipeButton extends StatefulWidget {
  const SwipeButton({
    Key key,
    this.thumb,
    this.content,
    BorderRadius borderRadius,
    this.initialPosition = SwipePosition.SwipeLeft,
    @required this.onChanged,
    this.swipeController,
    this.height = 56.0,
  })  : assert(initialPosition != null && onChanged != null && height != null),
        this.borderRadius = borderRadius ?? BorderRadius.zero,
        super(key: key);

  final Widget thumb;
  final Widget content;
  final BorderRadius borderRadius;
  final double height;
  final SwipePosition initialPosition;
  final ValueChanged<SwipePosition> onChanged;
  final SwipeController swipeController;

  @override
  SwipeButtonState createState() => SwipeButtonState();
}

class SwipeButtonState extends State<SwipeButton>
    with SingleTickerProviderStateMixin {
  final GlobalKey _containerKey = GlobalKey();
  final GlobalKey _positionedKey = GlobalKey();

  AnimationController _controller;
  Animation<double> _contentAnimation;
  Offset _start = Offset.zero;

  RenderBox get _positioned => _positionedKey.currentContext.findRenderObject();

  RenderBox get _container => _containerKey.currentContext.findRenderObject();

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(vsync: this);
    _contentAnimation = Tween<double>(begin: 1.0, end: 0.0)
        .animate(CurvedAnimation(parent: _controller, curve: Curves.easeOut));
    if (widget.initialPosition == SwipePosition.SwipeRight) {
      _controller.value = 1.0;
    }
    widget.swipeController?._attachController(_controller);
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    //final theme = Theme.of(context);
    return SizedBox(
      width: double.infinity,
      height: 100.0,
      child: Stack(
        key: _containerKey,
        children: <Widget>[
          DecoratedBox(
            decoration: BoxDecoration(
              color: Colors.deepPurpleAccent,
              borderRadius: new BorderRadius.all(new Radius.circular(50.0)),
            ),
            child: ClipRRect(
              clipper: _SwipeButtonClipper(
                animation: _controller,
                borderRadius: widget.borderRadius,
              ),
              borderRadius: widget.borderRadius,
              child: SizedBox.expand(
                child: Padding(
                  padding: EdgeInsets.only(left: widget.height),
                  child: FadeTransition(
                    opacity: _contentAnimation,
                    child: widget.content,
                  ),
                ),
              ),
            ),
          ),
          AnimatedBuilder(
            animation: _controller,
            builder: (BuildContext context, Widget child) {
              print(_controller.value);
              return Align(
                alignment: Alignment((_controller.value * 2.0) - 1.0, 0.0),
                child: child,
              );
            },
            child: GestureDetector(
              onHorizontalDragStart: _onDragStart,
              onHorizontalDragUpdate: _onDragUpdate,
              onHorizontalDragEnd: _onDragEnd,
              child: Container(
                key: _positionedKey,
                width: 100,
                height: 100,
                decoration: BoxDecoration(
                  shape: BoxShape.circle,
                  color: Colors.pinkAccent,
                ),
                child: widget.thumb,
              ),
            ),
          ),
        ],
      ),
    );
  }

  void _onDragStart(DragStartDetails details) {
    final pos = _positioned.globalToLocal(details.globalPosition);
    _start = Offset(pos.dx, 0.0);
    _controller.stop(canceled: true);
  }

  void _onDragUpdate(DragUpdateDetails details) {
    final pos = _container.globalToLocal(details.globalPosition) - _start;
    final extent = _container.size.width - _positioned.size.width;
    _controller.value = (pos.dx.clamp(0.0, extent) / extent);
  }

  void _onDragEnd(DragEndDetails details) {
    final extent = _container.size.width - _positioned.size.width;
    var fractionalVelocity = (details.primaryVelocity / extent).abs();
    if (fractionalVelocity < 0.5) {
      fractionalVelocity = 0.5;
    }
    SwipePosition result;
    double acceleration, velocity;
    if (_controller.value > 0.5) {
      acceleration = 0.5;
      velocity = fractionalVelocity;
      result = SwipePosition.SwipeRight;
    } else {
      acceleration = -0.5;
      velocity = -fractionalVelocity;
      result = SwipePosition.SwipeLeft;
    }
    final simulation = _SwipeSimulation(
      acceleration,
      _controller.value,
      1.0,
      velocity,
    );
    _controller.animateWith(simulation).then((_) {
      if (widget.onChanged != null) {
        widget.onChanged(result);
      }
    });
  }
}

class _SwipeSimulation extends GravitySimulation {
  _SwipeSimulation(
      double acceleration, double distance, double endDistance, double velocity)
      : super(acceleration, distance, endDistance, velocity);

  @override
  double x(double time) => super.x(time).clamp(0.0, 1.0);

  @override
  bool isDone(double time) {
    final _x = x(time).abs();
    return _x <= 0.0 || _x >= 1.0;
  }
}

class _SwipeButtonClipper extends CustomClipper<RRect> {
  const _SwipeButtonClipper({
    @required this.animation,
    @required this.borderRadius,
  })  : assert(animation != null && borderRadius != null),
        super(reclip: animation);

  final Animation<double> animation;
  final BorderRadius borderRadius;

  @override
  RRect getClip(Size size) {
    return borderRadius.toRRect(
      Rect.fromLTRB(
        size.width * animation.value,
        0.0,
        size.width,
        size.height,
      ),
    );
  }

  @override
  bool shouldReclip(_SwipeButtonClipper oldClipper) => true;
}


lib/swipe_button_demo.dart
library swipe_button;

import 'package:flutter/material.dart';
import 'package:swipe_button/swipe_button.dart';

class SwipeButtonDemo extends StatelessWidget {
  final String pageRoute;
  final String buttonTitle;

  const SwipeButtonDemo({Key key, this.pageRoute, this.buttonTitle})
      : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SizedBox(
        width: double.infinity,
        height: double.infinity,
        child: Stack(
          children: <Widget>[
            Container(
              padding: EdgeInsets.only(
                top: 50.0,
                left: 40.0,
              ),
              alignment: Alignment.topLeft,
              child: Text("Incident",
                  style: TextStyle(fontSize: 34.0, color: Colors.black87)),
            ),
            Container(
              padding: EdgeInsets.only(
                top: 50.0,
                right: 40.0,
              ),
              alignment: Alignment.topRight,
              child: Icon(
                Icons.print,
                color: Colors.amber,
                size: 50.0,
              ),
            ),
            Container(
              alignment: Alignment.center,
              child: Padding(
                padding: const EdgeInsets.all(24.0),
                child: SwipeButton(
                  thumb: Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: <Widget>[
                      Align(
                          widthFactor: 0.90,
                          child: Icon(
                            Icons.chevron_right,
                            size: 60.0,
                            color: Colors.white,
                          )),
                    ],
                  ),
                  content: Center(
                    child: Text(
                      buttonTitle,
                      style: TextStyle(color: Colors.white),
                    ),
                  ),
                  onChanged: (result) {
                    if (result == SwipePosition.SwipeRight) {
                      Navigator.pushNamed(context, pageRoute);
                    } else {}
                  },
                ),
              ),
            ),
            Container(
              margin: EdgeInsets.only(
                top: 250.0,
                left: 50.0,
              ),
              alignment: Alignment.centerLeft,
              child:
                  Text("Tips", style: TextStyle(fontWeight: FontWeight.bold)),
            ),
            Container(
              margin: EdgeInsets.only(
                top: 300.0,
                left: 50.0,
              ),
              alignment: Alignment.centerLeft,
              child: Text("Take 3 photograps for proof",
                  style: TextStyle(color: Colors.grey)),
            ),
            Container(
              margin: EdgeInsets.only(
                top: 350.0,
                left: 50.0,
              ),
              alignment: Alignment.centerLeft,
              child: Text("Obtainthe identy and personal details",
                  style: TextStyle(color: Colors.grey)),
            ),
            Container(
              margin: EdgeInsets.only(
                top: 400.0,
                left: 50.0,
              ),
              alignment: Alignment.centerLeft,
              child: Text("Specify the location",
                  style: TextStyle(color: Colors.grey)),
            ),
            Container(
              margin: EdgeInsets.only(
                top: 450.0,
                left: 50.0,
              ),
              alignment: Alignment.centerLeft,
              child: Text("Note Down The Vechicle number",
                  style: TextStyle(color: Colors.grey)),
            ),
          ],
        ),
      ),
    );
  }
}
